### PR TITLE
Provide dedicated page per authentication provider

### LINKED ISSUE #711: Provide dedicated page per authentication provider
As our offering for trigger authentication is growing, so does [our list of supported authentication providers](https://keda.sh/docs/2.6/concepts/authentication/).

We would love to align our approach in this area with our [scaler conceptual documentation](https://keda.sh/docs/2.6/concepts/scaling-deployments/) and [dedicated scaler overview](https://keda.sh/docs/2.6/scalers/).

With this change, there should be a list of supported authentication providers with a link to a dedicated page showing when it was introduced and how to use it.

### PR DESCRIPTION
**Pull request by cynthia-sg**
This PR splits the `Authentication Providers` documentation page content into
multiple pages. A dedicated page per authentication provider has been added,
and the main page now includes a list of the available providers with a link to
each page. This change has been applied to versions 2.7 and 2.8.

Fixes #711

Signed-off-by: Cintia Sanchez Garcia <cynthiasg@icloud.com>

### CODE REVIEW

File: content/docs/2.7/providers/azure-workload.md [hunk 1]
```diff
@@ -0,0 +1,27 @@
++++
+title = "Azure Workload Identity"
+layout = "provider"
```
Comment (tomkerkhove) [inline content/docs/2.7/providers/azure-workload.md:3]: This is a new one so we can leave it out - Thanks!

File: content/docs/2.7/concepts/authentication.md [hunk 2]
```diff
@@ -272,37 +273,4 @@ podIdentity:
   provider: none | azure | aws-eks | aws-kiam  # Optional. Default: none
 ```
 
-#### Azure Pod Identity
```
Comment (tomkerkhove) [inline content/docs/2.7/concepts/authentication.md:275]: All of the above are also identity providers: 
- Azure Key Vault secret(s) 
- Hashicorp Vault secret(s) 
- Secret(s)
- Environment variable(s)

Can we add those in individual files as well? If it helps; I'm happy to do so in another PR.

File: layouts/partials/nav.html [hunk 3]
```diff
@@ -35,13 +35,15 @@
       </li>
 
       {{ if $isHere }}
```
Comment (tomkerkhove) [inline layouts/partials/nav.html:37]: Can we add an entry next to "Scalers" in the main navigation?

Something like "Authentication Providers" or so. Thoughts on a better name @zroubalik @JorTurFer ?

**Review by tomkerkhove — COMMENTED** [review]
Did a first skim and I love it!

File: content/docs/2.7/providers/aws-eks.md [hunk 4]
```diff
@@ -0,0 +1,14 @@
++++
+title = "EKS Pod Identity Webhook for AWS"
+layout = "provider"
+weight = 3
```
Comment (chalin) [inline content/docs/2.7/providers/aws-eks.md:4]: I suggest that we drop the weights in this and other providers' files, and let the providers be listed alphabetically.

File: layouts/_default/provider.html [hunk 5]
```diff
@@ -0,0 +1,7 @@
+{{ define "title" }}
+Authentication provider | {{ .Title }}
+{{ end }}
+
+{{ define "main" }}
+{{ partial "article.html" . }}
+{{ end }}
```
Comment (chalin) [inline layouts/_default/provider.html:7]: As far as I can tell in my tests, this default layout isn't being picked up so you can drop it.

File: content/docs/2.8/providers/aws-eks.md [hunk 6]
```diff
@@ -0,0 +1,14 @@
++++
+title = "EKS Pod Identity Webhook for AWS"
+layout = "provider"
```
Comment (chalin) [inline content/docs/2.8/providers/aws-eks.md:3]: Drop `layout = "provider"` from this and other pages since the layout isn't being picked up. In any case, there is nothing special about how this page is rendered, right?

File: layouts/partials/content.html [hunk 7]
```diff
@@ -47,6 +48,10 @@ <h2 id="available-scalers">
         <script async src="https://artifacthub.io/artifacthub-widget.js"></script>
       </div>
       {{ end }}
+
+      {{ if $providersListVisible }}
+      {{ partial "providers.html" . }}
+      {{ end }}
```
Comment (chalin) [inline layouts/partials/content.html:54]: Instead of modifying this `contexnt.html` template, I'd recommend defining a shortcode and then using the shortcode in the relevant pages. Just move `providers.html` to be under shortcodes and then use it like this, e.g., in `authentication.md`:

```

Available authentication providers for KEDA:

{{< providers >}}
```

File: layouts/partials/nav.html [hunk 8]
```diff
@@ -35,13 +35,15 @@
...
+      {{ if ne .TableOfContents "<nav id=\"TableOfContents\"></nav>" }} <!-- not render ToC when is empty -->
```
Comment (chalin) [inline layouts/partials/nav.html:38]: I'm unsure why this extra `if` is necessary. I think that the `if` on the line below (line 39) serves the same purpose -- though the magic number "32" might be off and need to be adjusted.

File: layouts/partials/providers.html [hunk 9]
```diff
@@ -0,0 +1,15 @@
+{{ $here    := $.Page.RelPermalink }}
+{{ $providers := where site.RegularPages ".CurrentSection.Title" "Authentication Providers" }}
+{{ $version := index (split $.Page.File.Path "/") 1 }}
+<div class="buttons has-extra-top-margin">
+  {{ range $providers }}
+  {{ $isHere      := eq .RelPermalink $here }}
```
Comment (chalin) [inline layouts/partials/providers.html:6]: `$isHere` isn't relevant for providers AFAICT.

File: layouts/partials/nav.html (hunk 8 above)
Comment (chalin) [reply layouts/partials/nav.html:38]: Actually, 32 seems to work just fine according to my tests.

**Review by chalin — CHANGES_REQUESTED** [review]
Very nice work @cynthia-sg, thanks! See my inline comments.

**Comment by chalin** [comment]
@tomkerkhove - This PR presents the **Authentication Providers** in the left nav as a drop down. Do you expect this list of providers to grow like for that for Scalers? If so, consider not listing the providers via a drop down.

**Comment by chalin** [comment]
@cynthia-sg - I have some of the changes that I suggested via the review already done in my local repo, let me know if you'd like for me to push those up.

**Comment by tomkerkhove** [comment]
> @tomkerkhove - This PR presents the **Authentication Providers** in the left nav as a drop down. Do you expect this list of providers to grow like for that for Scalers? If so, consider not listing the providers via a drop down.

I don't think we'll have more than 10 or so long term but it does not have to be a dropdown perse; I would just like to have a shortcut here:
image

**Review by tomkerkhove — COMMENTED** [review]
Would you mind adding a quick section in the readme on how to add a new provider similar to what we have for scalers please?

File: layouts/partials/nav.html (hunk 3 above)
Comment (chalin) [reply layouts/partials/nav.html:37]: Something to consider: the Keda.sh top nav is getting crowded. It is already overflowing as you can see from this screenshot:

> image

Notice how the top nav is clipped on the right side of the screen.

Comment (tomkerkhove) [reply layouts/partials/nav.html:37]: That's OK; I'll create a seperate item as we can move Merch in Project

Comment (tomkerkhove) [reply layouts/partials/nav.html:37]: https://github.com/kedacore/keda-docs/issues/872

File: content/docs/2.7/providers/azure-workload.md (hunk 1 above)
Comment (cynthia-sg) [reply content/docs/2.7/providers/azure-workload.md]: Done!

File: content/docs/2.7/concepts/authentication.md (hunk 2 above)
Comment (cynthia-sg) [reply content/docs/2.7/concepts/authentication.md:275]: Done!

File: content/docs/2.7/providers/aws-eks.md (hunk 4 above)
Comment (cynthia-sg) [reply content/docs/2.7/providers/aws-eks.md]: Done!

File: layouts/_default/provider.html (hunk 5 above)
Comment (cynthia-sg) [reply layouts/_default/provider.html]: Removed!

File: content/docs/2.8/providers/aws-eks.md (hunk 6 above)
Comment (cynthia-sg) [reply content/docs/2.8/providers/aws-eks.md]: Not, nothing. Removed!

File: layouts/partials/content.html (hunk 7 above)
Comment (cynthia-sg) [reply layouts/partials/content.html]: Good point! Done!

**Comment by JorTurFer** [comment]
awesome work! One thing, maybe we should add a link or a note in the [Authentication section](https://deploy-preview-867--keda.netlify.app/docs/2.7/concepts/authentication) pointing to this because they are going to use there

File: layouts/partials/nav.html (hunk 8 above)
Comment (cynthia-sg) [reply layouts/partials/nav.html]: Yes, it's working using 32. 

I was thinking about asking to replace the new `if` for the other one. IMHO, the comparison with the empty nav code is more clear than counting characters. But I have removed it for now.

File: layouts/partials/providers.html (hunk 9 above)
Comment (cynthia-sg) [reply layouts/partials/providers.html]: Done!

**Comment by cynthia-sg** [comment]
> @cynthia-sg - I have some of the changes that I suggested via the review already done in my local repo, let me know if you'd like for me to push those up.

I’ve already addressed them, but thanks anyway!

**Comment by cynthia-sg** [comment]
> 



> 



> > @tomkerkhove - This PR presents the **Authentication Providers** in the left nav as a drop down. Do you expect this list of providers to grow like for that for Scalers? If so, consider not listing the providers via a drop down.
> 
> I don't think we'll have more than 10 or so long term but it does not have to be a dropdown perse; I would just like to have a shortcut here: image

The link for the `Authentication Providers` is there in fact. I added it for mobile version but it's hidden for desktop using the class `providers-link`. I agree with @chalin about the navbar is crowded and perhaps it's necessary to change this.

**Comment by cynthia-sg** [comment]
> Would you mind adding a quick section in the readme on how to add a new provider similar to what we have for scalers please?

Added! Please feel free to update the template for the provider.

**Comment by cynthia-sg** [comment]
> awesome work! One thing, maybe we should add a link or a note in the [Authentication section](https://deploy-preview-867--keda.netlify.app/docs/2.7/concepts/authentication) pointing to this because they are going to use there

Thanks! I'm not sure what you mean. Right now providers are listed at the end of the [Authentication section](https://deploy-preview-867--keda.netlify.app/docs/2.7/concepts/authentication/#available-authentication-providers-for-keda). Do you want a link/note in another part of the section too?

**Comment by JorTurFer** [comment]
> > awesome work! One thing, maybe we should add a link or a note in the [Authentication section](https://deploy-preview-867--keda.netlify.app/docs/2.7/concepts/authentication) pointing to this because they are going to use there
> 
> Thanks! I'm not sure what you mean. Right now providers are listed at the end of the [Authentication section](https://deploy-preview-867--keda.netlify.app/docs/2.7/concepts/authentication/#available-authentication-providers-for-keda). Do you want a link/note in another part of the section too?

oh, that's true...
Ignore me

File: README.md [hunk 10]
```diff
@@ -85,6 +85,20 @@ Make sure to update the following metadata fields:
 * `maintainer`
 * `description`
 
+## Adding provider documentation
```
Comment (chalin) [inline README.md:88]: @tomkerkhove et all: should this title have "authentication" somewhere in it? Maybe:

```suggestion
## Adding authentication-provider documentation
```

Or:

```suggestion
## Adding documentation for an authentication provider
```

File: content/docs/2.7/concepts/authentication.md (hunk 2 above)
Comment (chalin) [reply content/docs/2.7/concepts/authentication.md:275]: @tomkerkhove, you mentioned those 4 "identity providers", whcih @cynthia-sg has added. These now show up under https://deploy-preview-867--keda.netlify.app/docs/2.7/providers/, which has as a title "Authentication Providers". Maybe the section title should be changed to just **Providers**? (If we agree to the title change, then it should be changed everywhere -- i.e., including for the corresponding nav entry titles.)

If so, then the first paragraph in that page should be changed too, maybe:

> Available authentication and identity providers for KEDA:

WDYT?

File: content/docs/2.7/providers/azure-key-vault.md [hunk 11]
```diff
@@ -0,0 +1,38 @@
++++
+title = "Azure Key Vault secret(s)"
```
Comment (chalin) [inline content/docs/2.7/providers/azure-key-vault.md:2]: My suggestion for this and every other title containing "(s)" is to drop the parentheses and simply use the singular form (which seems to match the provider filename in most cases).

```suggestion
title = "Azure Key Vault secret"
```

The intent is that this would read as the _Azure Key Vault secret_ provider.

WDYT @tomkerkhove et all?

File: layouts/partials/content.html [hunk 12]
```diff
@@ -50,4 +50,3 @@ <h2 id="available-scalers">
     </div>
   </div>
 </section>
-
```
Comment (chalin) [inline layouts/partials/content.html:53]: Nit: consider undoing this whitespace change since it's the only change in the file.

**Review by chalin — APPROVED** [review]
See inline for comments about terminology, which might affect titles. Otherwise, LGTM!

File: README.md (hunk 10 above)
Comment (tomkerkhove) [reply README.md:88]: Yes, I'd suggest to go with this to follow the scaler approach
```suggestion
## Writing documentation for a new authentication provider
```

File: content/docs/2.7/providers/azure-key-vault.md (hunk 11 above)
Comment (tomkerkhove) [reply content/docs/2.7/providers/azure-key-vault.md:2]: Fine by me, I don't have a preference

File: content/docs/2.7/concepts/authentication.md (hunk 2 above)
Comment (tomkerkhove) [reply content/docs/2.7/concepts/authentication.md:275]: I would not rename them to "providers" because it's not clear what the providers are for - Are they autoscaling providers or authentication providers?

That's why I would stick with "authentication providers" and also include this in the URL (be it "auth-providers", "authentication-providers", "providers/auth/xxx").

What do you think @zroubalik?

**Comment by tomkerkhove** [comment]
> > 
> 
> > 
> 
> > > @tomkerkhove - This PR presents the **Authentication Providers** in the left nav as a drop down. Do you expect this list of providers to grow like for that for Scalers? If so, consider not listing the providers via a drop down.
> > 
> > 
> > I don't think we'll have more than 10 or so long term but it does not have to be a dropdown perse; I would just like to have a shortcut here: image
> 
> The link for the `Authentication Providers` is there in fact. I added it for mobile version but it's hidden for desktop using the class `providers-link`. I agree with @chalin about the navbar is crowded and perhaps it's necessary to change this.

Fine by me, but then it should be in this drop-down at least because it's hidden atm:
image

File: content/docs/2.7/concepts/authentication.md [hunk 13]
```diff
@@ -176,91 +176,6 @@ spec:
...
-    type: AzurePublicCloud | AzureUSGovernmentCloud | AzureChinaCloud | AzureGermanCloud | Private # Required.
-    keyVaultResourceURL: {key-vault-resource-url-for-cloud}           # Required when type = Private.
-    activeDirectoryEndpoint: {active-directory-endpoint-for-cloud}    # Required when type = Private.
-  secrets:                                              # Required.
-  - parameter: {param-name-used-for-auth}               # Required.
-    name: {key-vault-secret-name}                       # Required.
-    version: {key-vault-secret-version}                 # Optional.
-```
-
 ### Pod Authentication Providers
```
Comment (tomkerkhove) [inline content/docs/2.7/concepts/authentication.md:179]: It's OK to drop this section as well given this is now in "Azure pod identity" and the related dedicated pages.

File: content/docs/2.7/concepts/authentication.md [hunk 14]
```diff
@@ -176,91 +176,6 @@ spec:
```
Comment (tomkerkhove) [inline content/docs/2.7/concepts/authentication.md:176]: I would rename the "Authentication Parameters" header to the existing "Available authentication providers for KEDA" below so that it is the following:

> ## Available authentication providers for KEDA
> 
> Authentication parameters can be retrieved from a variety of authentication providers in KEDA: 
> 
> -list auth providers-

**Review by tomkerkhove — APPROVED** [review]
LGTM, thanks a ton for this!

Here are my last remarks along with some comments above; for the rest things should be OK!

**Comment by tomkerkhove** [comment]
Sorry for all the comments but I was wondering if you are up for updating this PR @cynthia-sg? I appreciate the big help here, this is tremendous help!

**Comment by cynthia-sg** [comment]
> Sorry for all the comments but I was wondering if you are up for updating this PR @cynthia-sg? I appreciate the big help here, this is tremendous help!

Hi @tomkerkhove,

Yes, I will take care of it soon, leave it with me. I’ve been a few days off on vacation :)

**Comment by tomkerkhove** [comment]
No worries at all and hope you've enjoyed your time off!

File: content/docs/2.7/providers/azure-key-vault.md (hunk 11 above)
Comment (cynthia-sg) [reply content/docs/2.7/providers/azure-key-vault.md]: Done! 👍

File: content/docs/2.7/concepts/authentication.md (hunk 13 above)
Comment (cynthia-sg) [reply content/docs/2.7/concepts/authentication.md]: Done!

File: content/docs/2.7/concepts/authentication.md (hunk 14 above)
Comment (cynthia-sg) [reply content/docs/2.7/concepts/authentication.md:176]: Done! 👍

File: content/docs/2.7/concepts/authentication.md (hunk 2 above)
Comment (cynthia-sg) [reply content/docs/2.7/concepts/authentication.md:275]: I'm using right now `authentication-providers`, wdyt?

> > > > @tomkerkhove - This PR presents the **Authentication Providers** in the left nav as a drop down. Do you expect this list of providers to grow like for that for Scalers? If so, consider not listing the providers via a drop down.
> > > 
> > > 
> > > I don't think we'll have more than 10 or so long term but it does not have to be a dropdown perse; I would just like to have a shortcut here: image
> > 
> > 
> > The link for the `Authentication Providers` is there in fact. I added it for mobile version but it's hidden for desktop using the class `providers-link`. I agree with @chalin about the navbar is crowded and perhaps it's necessary to change this.
> 
> Fine by me, but then it should be in this drop-down at least because it's hidden atm: image

I have added the link to the dropdown and removed the previous one.

keda-link

File: README.md (hunk 10 above)
Comment (cynthia-sg) [reply README.md]: Done! 👍

File: content/docs/2.7/concepts/authentication.md (hunk 2 above)
Comment (cynthia-sg) [reply content/docs/2.7/concepts/authentication.md:275]: > I would not rename them to "providers" because it's not clear what the providers are for - Are they autoscaling providers or authentication providers?
> 
> That's why I would stick with "authentication providers" and also include this in the URL (be it "auth-providers", "authentication-providers", "providers/auth/xxx").
> 
> What do you think @zroubalik?

I'm used `authentication-providers` as you suggested @tomkerkhove.

File: layouts/partials/content.html (hunk 12 above)
Comment (cynthia-sg) [reply layouts/partials/content.html]: Done!

**Comment by tomkerkhove** [comment]
Can we add a link on the landing page as well?

**Review by tomkerkhove — APPROVED** [review]
Added one comment about landing page but love it, thanks a ton!

**Comment by tomkerkhove** [comment]
Still not sold on the side nav missing a link per provider though be we can continue that in a seperate PR/issue 🙂

**Comment by cynthia-sg** [comment]
> Can we add a link on the landing page as well?
> 
> 

Added! 👍 
keda-home

**Comment by tomkerkhove** [comment]
Thanks a ton!

**Comment by tomkerkhove** [comment]
I love it.

**Comment by cynthia-sg** [comment]
> I love it.

🎉
